File: gates/chunking-parity/probe.py
#!/usr/bin/env python3
"""gates/chunking-parity/probe.py — the `python` arm of the chunking-parity gate.

The `typescript` twin, and deliberately the same shape: read the shared corpus, run
§3.6 FastCDC/NC2 through THE STAGED PACKAGE imported by name, print one JSON line.

Three numbers, each answering a different question:
  boundary_count   -- did the three ports cut the same NUMBER of chunks?
  boundary_digest  -- did they cut at the same OFFSETS? (a count can match by accident)
  blob_hash        -- do the resulting §2.1 entities hash identically? This is the one
                      that matters: it is the deduplication identity, and two peers that
                      disagree here silently fail to dedup with each other while every
                      blob still reassembles.
"""

from __future__ import annotations

import hashlib
import json
import sys

from entity_content import cdc_boundaries, create_blob_cdc


def main() -> int:
    if len(sys.argv) != 3:
        print("usage: probe.py <corpus-path> <target-size>", file=sys.stderr)
        return 2
    corpus_path, target_size = sys.argv[1], int(sys.argv[2])

    with open(corpus_path, "rb") as fh:
        data = fh.read()

    bounds = cdc_boundaries(data, target_size)
    blob = create_blob_cdc(data, target_size)

    print(json.dumps({
        "port": "python",
        "corpus_bytes": len(data),
        "target_size": target_size,
        "boundary_count": len(bounds),
        "boundary_digest": hashlib.sha256(",".join(map(str, bounds)).encode()).hexdigest(),
        "blob_hash": bytes(blob.blob.hash).hex(),
        "chunk_count": len(blob.chunks),
        "first_chunk_hash": bytes(blob.chunks[0].hash).hex(),
        "last_chunk_hash": bytes(blob.chunks[-1].hash).hex(),
    }))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: languages/typescript/gates/chunking-parity/probe.mjs
// gates/chunking-parity/probe.mjs — the `typescript` arm of the chunking-parity gate.
//
// Reads the shared corpus, runs §3.6 FastCDC/NC2 through THE STAGED PACKAGE (resolved
// by name through its `exports` map, not by a path into the source tree), and prints one
// JSON line. `compare.py` diffs the three.
//
// Three numbers, each answering a different question:
//   boundary_count   — did the three ports cut the same NUMBER of chunks?
//   boundary_digest  — did they cut at the same OFFSETS? (a count can match by accident)
//   blob_hash        — do the resulting §2.1 entities hash identically? This is the one
//                      that matters: it is the deduplication identity, and two peers that
//                      disagree here silently fail to dedup with each other while every
//                      blob still reassembles.

import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";

const corpusPath = process.argv[2];
const targetSize = Number(process.argv[3]);
if (!corpusPath || !Number.isInteger(targetSize)) {
  console.error("usage: probe.mjs <corpus-path> <target-size>");
  process.exit(2);
}

const { cdcBoundaries, createBlobCdc } = await import("@entity-core/extension-content");

const data = new Uint8Array(readFileSync(corpusPath));
const bounds = cdcBoundaries(data, targetSize);
const blob = createBlobCdc(data, targetSize);

// `contentHash`, not `hash`. The three peers name this field three ways -- `contentHash`
// here, `hash` on both `python` and `rust` -- and the first draft of this probe reached
// for `.hash`, got `undefined`, and threw inside `Buffer.from`. It threw, which is the
// good outcome: a probe that had silently hexed `undefined` to "" would have reported
// three identical empty blob hashes and read as perfect parity.
const hex = (bytes) => {
  if (!(bytes instanceof Uint8Array)) {
    throw new TypeError(`expected a Uint8Array content hash, got ${typeof bytes}`);
  }
  return Buffer.from(bytes).toString("hex");
};

console.log(
  JSON.stringify({
    port: "typescript",
    corpus_bytes: data.length,
    target_size: targetSize,
    boundary_count: bounds.length,
    boundary_digest: createHash("sha256").update(bounds.join(",")).digest("hex"),
    blob_hash: hex(blob.blob.contentHash),
    chunk_count: blob.chunks.length,
    first_chunk_hash: hex(blob.chunks[0].contentHash),
    last_chunk_hash: hex(blob.chunks[blob.chunks.length - 1].contentHash),
  }),
);
